pkg/build/controller/common/util.go

```diff
@@ -42,6 +47,83 @@ func HandleBuildCompletion(build *buildapi.Build, runPolicies []policy.RunPolicy
 	if err := runPolicy.OnComplete(build); err != nil {
 		utilruntime.HandleError(fmt.Errorf("failed to run policy on completed build %s/%s: %v", build.Namespace, build.Name, err))
 	}
+	if err := HandleBuildPruning(build, buildLister, buildConfigGetter, buildDeleter); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to prune old builds %s/%s: %v", build.Namespace, build.Name, err))
+	}
+}
+
+func isCompletedBuild(build buildapi.Build) bool {
+	return build.Status.Phase == buildapi.BuildPhaseComplete
+}
+
+func isFailedBuild(build buildapi.Build) bool {
+	return build.Status.Phase == buildapi.BuildPhaseFailed
+}
+
+type ByStartTimestamp []buildapi.Build
+
+func (b ByStartTimestamp) Len() int {
+	return len(b)
+}
+
+func (b ByStartTimestamp) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+func (b ByStartTimestamp) Less(i, j int) bool {
+	return !b[j].Status.StartTimestamp.Time.After(b[i].Status.StartTimestamp.Time)
+}
+
+// HandleBuildPruning handles the deletion of old successful and failed builds
+// based on settings in the BuildConfig.
+func HandleBuildPruning(build *buildapi.Build, buildLister buildclient.BuildLister, buildConfigGetter buildclient.BuildConfigGetter, buildDeleter pruner.BuildDeleter) error {
+	bcName := buildutil.ConfigNameForBuild(build)
+
+	buildConfig, err := buildConfigGetter.Get(build.Namespace, bcName)
+	if err != nil {
+		return err
+	}
+
+	if buildConfig.Spec.FailedBuildsHistoryLimit == nil && buildConfig.Spec.SuccessfulBuildsHistoryLimit == nil {
+		return nil
+	}
+
+	successfulBuilds, err := buildutil.BuildConfigBuilds(buildLister, build.Namespace, bcName, isCompletedBuild)
+	sort.Sort(ByStartTimestamp(successfulBuilds.Items))
+
+	failedBuilds, err := buildutil.BuildConfigBuilds(buildLister, build.Namespace, bcName, isFailedBuild)
+	sort.Sort(ByStartTimestamp(failedBuilds.Items))
+
+	var buildsToDelete []buildapi.Build
+
+	if buildConfig.Spec.SuccessfulBuildsHistoryLimit != nil {
+		successfulBuildsHistoryLimit := int(*buildConfig.Spec.SuccessfulBuildsHistoryLimit)
+		glog.V(4).Infof("Preparing to prune %v old successful builds.", (len(successfulBuilds.Items) - successfulBuildsHistoryLimit))
+		if successfulBuildsHistoryLimit == 0 && len(successfulBuilds.Items) > 0 {
```

Comment:
Yep, that's a leftover from the way it was being done before, fixed in next push